ArduCopter/mode_systemid.cpp

```diff
@@ -0,0 +1,361 @@
+                    break;
+                case INPUT_PITCH:
+                    target_pitch += waveformSample*100.0f;
+                    break;
+                case INPUT_YAW:
+                    target_yaw_rate += waveformSample*100.0f;
+                    break;
+                case RECOVER_ROLL:
+                    target_roll += waveformSample*100.0f;
+                    attitude_control->bf_feedforward(false);
+                    break;
+                case RECOVER_PITCH:
+                    target_pitch += waveformSample*100.0f;
+                    attitude_control->bf_feedforward(false);
+                    break;
+                case RECOVER_YAW:
+                    target_yaw_rate += waveformSample*100.0f;
+                    attitude_control->bf_feedforward(false);
+                    break;
+                case RATE_ROLL:
+                    attitude_control->rate_bf_roll_sysid(radians(waveformSample));
+                    break;
+                case RATE_PITCH:
+                    attitude_control->rate_bf_pitch_sysid(radians(waveformSample));
+                    break;
+                case RATE_YAW:
+                    attitude_control->rate_bf_yaw_sysid(radians(waveformSample));
+                    break;
+                case MIX_ROLL:
+                    attitude_control->actuator_roll_sysid(waveformSample);
+                    break;
+                case MIX_PITCH:
+                    attitude_control->actuator_pitch_sysid(waveformSample);
+                    break;
+                case MIX_YAW:
+                    attitude_control->actuator_yaw_sysid(waveformSample);
+                    break;
+                case MIX_THROTTLE:
+                    pilot_throttle_scaled += waveformSample;
+                    break;
+            }
+            break;
+    }
+
+    // call attitude controller
+    attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(target_roll, target_pitch, target_yaw_rate);
+
+    // body-frame rate controller is run directly from 100hz loop
+
+    // output pilot's throttle
+    if (copter.is_tradheli()) {
+        attitude_control->set_throttle_out(pilot_throttle_scaled, false, g.throttle_filt);
+    } else {
+        attitude_control->set_throttle_out(pilot_throttle_scaled, true, g.throttle_filt);
+    }
+
+    if(log_subsample <= 0){
+        log_data();
+        if(copter.should_log(MASK_LOG_ATTITUDE_FAST) && copter.should_log(MASK_LOG_ATTITUDE_MED)) {
+            log_subsample = 1;
+        } else if(copter.should_log(MASK_LOG_ATTITUDE_FAST)){
+            log_subsample = 2;
+        } else if(copter.should_log(MASK_LOG_ATTITUDE_MED)){
+            log_subsample = 4;
+        } else {
+            log_subsample = 8;
+        }
+    }
+    log_subsample -= 1;
+
+    log_data();
+}
+
+// init_test - initialises the test
+void ModeSystemId::log_data()
+{
+    int8_t index = copter.EKF2.getPrimaryCoreIMUIndex();
```

Comment:
we probably shouldn't hard-code in this dependency on the EKF2 because some (like Solo users) regularly use the EKF3.  Instead use the AP_AHRS_NavEKF::get_primary_IMU_index() function which should return the same thing.

Reply:
I think that is private but I have used the gyro and accel version. So I think I got this right.